File: leetcode/binary-tree-level-order-traversal.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
 public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> res;

        if (root) {
            TreeNode *node;
            int n;
            deque<pair<TreeNode *, int>> queue;

            queue.push_back({root, 0});

            while (!queue.empty()) {
                tie(node, n) = queue.front();
                queue.pop_front();

                if (res.size() <= n) {
                    res.push_back(vector<int>());
                }

                res[n].push_back(node->val);

                if (node->left)
                    queue.push_back({node->left, n + 1});
                if (node->right)
                    queue.push_back({node->right, n + 1});
            }
        }

        return res;
    }
};

File: leetcode/binary-tree-level-order-traversal.py
from collections import deque
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue = deque()

        if root:
            queue.append((root, 0))

        res = []

        while queue:
            node, level = queue.popleft()

            if len(res) < level + 1:
                res.append([])

            res[level].append(node.val)

            if node.left:
                queue.append((node.left, level + 1))
            if node.right:
                queue.append((node.right, level + 1))

        return res
